PLAT-1220: Add sortable component (#211)
reciprocity/zen-ui: src/components/zen-sortable/zen-sortable.stories.mdx

import { html } from 'lit-html';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { waitForElement } from '../../../.storybook/helpers/utils';
import { getArgTypesAndArgs, getDefaultArgs, getComponentData, spreadArgs } from '../../../.storybook/helpers/argTypes';
const { argTypes } = getArgTypesAndArgs('zen-sortable');

<Meta title="In progress/Sortable/Sortable" component="zen-sortable" argTypes={argTypes} />

export const StorySortable = () => {
  waitForElement('#sortable', element => {
    element.addEventListener('onChange', ev => {
      console.log('event: ', ev);
    });
  });
  return html/*html*/ ` <style>
      .container {
        width: 100%;
      }
    </style>

>     <sb-zen-sortable id="sortable" padding="lg" spacing="lg" style="width:230px;border:1px solid lightgray">

Author: No need. Only removed it because it defaults to none

      <sb-zen-sortable-item data-id="1">

>         <sb-zen-space class="container" fill horizontal-align="space-between" vertical-align="center">

Reviewer: Remove all these `zen-space` wrapper of each item after you add `zen-space` to item component.

Author: Removed

          <sb-zen-text>Title</sb-zen-text>
          <sb-zen-toggle></sb-zen-toggle>
        </sb-zen-space>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="2">
        <sb-zen-space class="container" fill horizontal-align="space-between" vertical-align="center">
          <sb-zen-text>Description</sb-zen-text>
          <sb-zen-toggle></sb-zen-toggle>
        </sb-zen-space>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="3">
        <sb-zen-space class="container" fill horizontal-align="space-between" vertical-align="center">
          <sb-zen-text>Status</sb-zen-text>
          <sb-zen-toggle></sb-zen-toggle>
        </sb-zen-space>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="4">
        <sb-zen-space class="container" fill horizontal-align="space-between" vertical-align="center">
          <sb-zen-text>Manager</sb-zen-text>
          <sb-zen-toggle></sb-zen-toggle>
        </sb-zen-space>
      </sb-zen-sortable-item>
    </sb-zen-sortable>`;
};

export const StoryWithControls = args => {

>   waitForElement('#controls', element => {

Author: Done. and event is already documented.

>   waitForElement('#controls', element => {

Reviewer: It's still not showing up in the stories. Change `<docs-table>` in your story to `<sb-docs-table>`

    element.addEventListener('onChange', ev => {
      console.log('event: ', ev);
    });
  });
  return html/*html*/ ` <sb-zen-sortable
    id="controls"
    ...="${spreadArgs(args, argTypes)}"
    style="border:1px solid lightgray"
  >
    <sb-zen-sortable-item data-id="1">
      <sb-zen-text>Item 1</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="2">
      <sb-zen-text>Item 2</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="3">
      <sb-zen-text>Item 3</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="4">
      <sb-zen-text>Item 4</sb-zen-text>
    </sb-zen-sortable-item>
  </sb-zen-sortable>`;
};

# Zen Sortable

> Slotted content becomes draggable. On change position it returns an "onChange" custom event that contains list of "data-id" attributes value.

Reviewer: Change to:
Slotted `zen-sortable-items` becomes draggable. Each draggable item should have unique attribute defined by prop `key`. On item drag&drop `drop` event is triggered. It contains list of item keys that resembles new order.

Author: Done. But i dont think its correct. Everything becomes slotted not only item. The uniq key is named data-id.

Reviewer 2: - Don't use `key` as an attribute, even more so for a component that will be repeated: `key` is used internally by React on loops that generate components. Keep `data-id`.
- I agree with the part about `Slotted `zen-sortable-items` because even if you add any other element, when you instantiate `Sortable`, you give it `.handle` as the handler, so whatever they put inside should have that class.

<br />

## Component previews

<br />

### 1. Sortable items

A composition of `ZenSortable` component and fragment `ZenSortableItem` components with custom default slot content.

<Canvas>
  <Story name="Sortable">{StorySortable.bind({})}</Story>
</Canvas>

## Properties

<Canvas>
  <Story
    name="default"
    args={{
      ...getDefaultArgs(argTypes),

>       padding: 'lg',

Author: Removed

import { html } from 'lit-html';
import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks';
import { logEvents } from '../../../.storybook/helpers/log-events';
import { getArgTypesAndArgs, getDefaultArgs, getComponentData, spreadArgs } from '../../../.storybook/helpers/argTypes';
const { argTypes } = getArgTypesAndArgs('zen-sortable');

<Meta title="Lists/Sortable/Sortable" component="zen-sortable" argTypes={argTypes} />

export const StorySortable = () => {
  return html/*html*/ `

>     <sb-zen-sortable id="sortable" spacing="sm" padding="sm" style="width:230px;border:1px solid lightgray;">
>       <sb-zen-sortable-item padding="md" data-id="1">
>           <sb-zen-text>Title</sb-zen-text>
>           <sb-zen-toggle></sb-zen-toggle>
>       </sb-zen-sortable-item>
>       <sb-zen-sortable-item padding="md" data-id="2">
>           <sb-zen-text>Description</sb-zen-text>
>           <sb-zen-toggle></sb-zen-toggle>
>       </sb-zen-sortable-item>
>       <sb-zen-sortable-item padding="md" data-id="3">
>           <sb-zen-text>Status</sb-zen-text>
>           <sb-zen-toggle style=></sb-zen-toggle>
>       </sb-zen-sortable-item>
>       <sb-zen-sortable-item padding="md" data-id="4">
>           <sb-zen-text>Manager</sb-zen-text>
>           <sb-zen-toggle></sb-zen-toggle>
>       </sb-zen-sortable-item>
>     </sb-zen-sortable>

Reviewer: ```suggestion
    <sb-zen-sortable id="sortable" style="width:230px;border:1px solid lightgray;">
      <sb-zen-sortable-item padding="md" data-id="1">
          <sb-zen-text>Title</sb-zen-text>
          <sb-zen-toggle class="space-before"></sb-zen-toggle>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="2">
          <sb-zen-text>Description</sb-zen-text>
          <sb-zen-toggle class="space-before"></sb-zen-toggle>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="3">
          <sb-zen-text>Status</sb-zen-text>
          <sb-zen-toggle class="space-before"></sb-zen-toggle>
      </sb-zen-sortable-item>
      <sb-zen-sortable-item data-id="4">
          <sb-zen-text>Manager</sb-zen-text>
          <sb-zen-toggle class="space-before"></sb-zen-toggle>
      </sb-zen-sortable-item>
    </sb-zen-sortable>
```

Reviewer: Make story match the designs.

Author: Done

  ${logEvents('#sortable', getComponentData('zen-sortable'))}`;
};

export const StoryWithControls = args => {
  return html/*html*/ ` <sb-zen-sortable
    id="controls"
    ...="${spreadArgs(args, argTypes)}"
    style="border:1px solid lightgray;"
  >
    <sb-zen-sortable-item data-id="1">
      <sb-zen-text>Item 1</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="2">
      <sb-zen-text>Item 2</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="3">
      <sb-zen-text>Item 3</sb-zen-text>
    </sb-zen-sortable-item>
    <sb-zen-sortable-item data-id="4">
      <sb-zen-text>Item 4</sb-zen-text>
    </sb-zen-sortable-item>
  </sb-zen-sortable>
  ${logEvents('#controls', getComponentData('zen-sortable'))}`;
};

# Zen Sortable

> Slotted `zen-sortable-items` becomes draggable. Each draggable item should have unique attribute defined by prop `key`. On item drag&drop `drop` event is triggered. It contains list of item keys that resembles new order.

Reviewer: ```suggestion
Slotted elements becomes draggable. It's advised to use `zen-sortable-items` for slotted elements.  Each draggable item should have unique `data-id` attribute. On item drag&drop `drop` event is triggered. It contains list of `data-id`s that resembles new order.
```

Reviewer: Changing it a bit, so it matches suggestions you've made in the comments. Feel free to change.

Author: Done

<br />

## Component previews

<br />

### 1. Sortable items

A composition of `ZenSortable` component and fragment `ZenSortableItem` components with custom default slot content.

<Canvas>
  <Story name="Sortable">{StorySortable.bind({})}</Story>
</Canvas>

## Properties

<Canvas>
  <Story
    name="default"
    args={{
      ...getDefaultArgs(argTypes),
    }}
  >
    {StoryWithControls.bind({})}
  </Story>
</Canvas>

<ArgsTable story="default" />

> <docs-table docs={JSON.stringify(getComponentData('zen-sortable'))}></docs-table>

Reviewer: ```suggestion
<sb-docs-table docs={JSON.stringify(getComponentData('zen-sortable'))}></sb-docs-table>
```

Reviewer: docs-table is our component and needs to be prefixed. This is why it didn't show documented events.

Author: Fixed
